js/src/functions/stream.ts

```diff
@@ -0,0 +1,204 @@
+import { callEventSchema } from "@braintrust/core/typespecs";
+import {
+  createParser,
+  EventSourceParser,
+  ParsedEvent,
+  ReconnectInterval,
+} from "eventsource-parser";
+
+/**
+ * A chunk of data from a Braintrust stream. Each chunk type matches
+ * an SSE event type.
+ */
+export type BraintrustStreamChunk =
+  | {
+      type: "text_delta";
+      data: string;
+    }
+  | {
+      type: "json_delta";
+      data: string;
+    };
+
+/**
+ * A Braintrust stream. This is a wrapper around a ReadableStream of `BraintrustStreamChunk`,
+ * with some utility methods to make them easy to log and convert into various formats.
+ */
+export class BraintrustStream {
+  private stream: ReadableStream<BraintrustStreamChunk>;
+  private memoizedFinalValue: Promise<unknown> | undefined;
+
+  constructor(baseStream: ReadableStream<Uint8Array>);
+  constructor(stream: ReadableStream<string>);
+  constructor(stream: ReadableStream<BraintrustStreamChunk>);
+  constructor(
+    baseStream:
+      | ReadableStream<Uint8Array>
+      | ReadableStream<string>
+      | ReadableStream<BraintrustStreamChunk>,
+  ) {
+    this.stream = baseStream.pipeThrough(btStreamParser());
+  }
+
+  /**
+   * Copy the stream. This returns a new stream that shares the same underlying
+   * stream (via `tee`). Since streams are consumed in Javascript, use `copy()` if you
+   * need to use the stream multiple times.
+   *
+   * @returns A new stream that you can independently consume.
+   */
+  public copy(): BraintrustStream {
+    // Once a stream is tee'd, it is essentially consumed, so we need to replace our own
+    // copy of it.
+    const [newStream, copyStream] = this.stream.tee();
+    this.stream = copyStream;
+    return new BraintrustStream(newStream);
+  }
+
+  /**
+   * Get the underlying ReadableStream.
+   *
+   * @returns The underlying ReadableStream<BraintrustStreamChunk>.
+   */
+  public toReadableStream(): ReadableStream<BraintrustStreamChunk> {
+    return this.stream;
+  }
+
+  /**
+   * Get the final value of the stream. This will return a promise that resolves
```

Comment:
The term "final value" is ambiguous to me. I read a bunch of the PR thinking that this meant the last chunk returned by the stream, but it's actually "the concatenation of all stream chunks". Maybe clarify this somewhere, and/or call this something like `fullStreamContents`?

Reply:
I don't think so. Final is what Vercel uses

Reply:
Ok but i think it should be clarified in the docs.